File: src/decoders/tab.ts
import type { Period, Tab } from "~/models";

export const decodeTab = (tab: any, instancePeriods: Period[]): Tab => {
  const find = (id: string) => instancePeriods.find((p2) => p2.id === id);

  const defaultPeriod = tab.periodeParDefaut && find(tab.periodeParDefaut.V.N);
  const periods = tab.listePeriodes.V.map((p1: any) => find(p1.N)).filter(Boolean);

  return {
    defaultPeriod,
    location: tab.G,
    periods
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/tab.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.Period
import ink.literate.pawnote.models.Tab
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.json.*

fun decodeTab(tab: JsonObject, instancePeriods: List<Period>): Tab {
  val find: (id: String) -> Period? = { id -> instancePeriods.find { it.id == id } }

  val defaultPeriod =
      if (tab["periodeParDefaut"] != null)
          find(tab["periodeParDefaut"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content)
      else null
  val periods =
      tab["listePeriodes"]!!.jsonObject["V"]!!.jsonArray.mapNotNull {
        if (it.jsonObject["N"] != null) find(it.jsonObject["N"]!!.jsonPrimitive.content) else null
      }

  return Tab(
      defaultPeriod = defaultPeriod,
      location = TabLocation.fromInt(tab["G"]!!.jsonPrimitive.int)!!,
      periods = periods)
}
